Regression: `TRANSIENT_LOCAL` durability does not deliver `std_msgs/msg/UInt8MultiArray` to late-joining subscribers (works on Jazzy/Kilted, fails on Lyrical)

### Generated by Generative AI

Sonnet-5

### Operating System:

Ubuntu 26.04 LTS (Resolute Raccoon)

### ROS version or commit hash:

lyrical

### RMW implementation (if applicable):

rmw_fastrtps_cpp

### RMW Configuration (if applicable):

_No response_

### Client library (if applicable):

rclcpp, rclpy

### 'ros2 doctor --report' output

<details><summary>ros2 doctor --report</summary>

```console
   ACTION LIST
action                 : none
action server count    : 0
action client count    : 0

   ROS ENVIRONMENT
ROS environment variables        : ROS_AUTOMATIC_DISCOVERY_RANGE=SUBNET, ROS_DISTRO=lyrical
rcutils environment variables    : 
rmw environment variables        : 

   NETWORK CONFIGURATION
inet         : 127.0.0.1
inet4        : ['127.0.0.1']
inet6        : ['::1']
netmask      : 255.0.0.0
device       : lo
flags        : UP,LOOPBACK,RUNNING
mtu          : 65536
inet         : 172.17.0.2
inet4        : ['172.17.0.2']
ether        : c2:a7:e2:8b:a0:a1
netmask      : 255.255.0.0
device       : eth0
flags        : UP,BROADCAST,RUNNING,MULTICAST
mtu          : 1500
broadcast    : 172.17.255.255

   PACKAGE VERSIONS
uncrustify_vendor                         : latest=3.2.0, local=3.2.0
ament_cmake_mypy                          : latest=0.20.6, local=0.20.6
ament_cmake_ros_core                      : latest=0.15.8, local=0.15.8
rosidl_runtime_py                         : latest=0.15.2, local=0.15.2
ament_cmake_gtest                         : latest=2.8.7, local=2.8.7
ament_index_python                        : latest=1.13.3, local=1.13.3
ament_mypy                                : latest=0.20.6, local=0.20.6
ros2component                             : latest=0.40.7, local=0.40.7
pluginlib                                 : latest=5.8.4, local=5.8.4
ament_cmake_python                        : latest=2.8.7, local=2.8.7
rmw_test_fixture                          : latest=0.15.8, local=0.15.8
ament_lint_common                         : latest=0.20.6, local=0.20.6
launch_ros                                : latest=0.29.9, local=0.29.8
rcl                                       : latest=10.4.4, local=10.4.4
ros2cli_common_extensions                 : latest=0.5.2, local=0.5.2
sensor_msgs                               : latest=5.9.2, local=5.9.2
ament_package                             : latest=0.18.3, local=0.18.3
class_loader                              : latest=2.9.4, local=2.9.4
rosidl_generator_py                       : latest=0.27.2, local=0.27.2
nav_msgs                                  : latest=5.9.2, local=5.9.2
ros2param                                 : latest=0.40.7, local=0.40.7
ros2topic                                 : latest=0.40.7, local=0.40.7
ament_cmake_version                       : latest=2.8.7, local=2.8.7
launch_testing_ros                        : latest=0.29.9, local=0.29.8
ros_core                                  : latest=0.13.0, local=0.13.0
ament_xmllint                             : latest=0.20.6, local=0.20.6
ament_cmake_core                          : latest=2.8.7, local=2.8.7
action_msgs                               : latest=2.4.5, local=2.4.5
libstatistics_collector                   : latest=2.1.2, local=2.1.2
ament_cmake_uncrustify                    : latest=0.20.6, local=0.20.6
ament_lint_cmake                          : latest=0.20.6, local=0.20.6
ament_cmake_target_dependencies           : latest=2.8.7, local=2.8.7
rosidl_generator_rs                       : latest=0.4.12, local=0.4.12
ament_cmake_export_libraries              : latest=2.8.7, local=2.8.7
rosidl_cmake                              : latest=5.2.1, local=5.2.1
rcutils                                   : latest=7.1.2, local=7.1.1
rmw                                       : latest=7.10.1, local=7.10.1
statistics_msgs                           : latest=2.4.5, local=2.4.5
ament_cmake_test                          : latest=2.8.7, local=2.8.7
rclcpp_action                             : latest=32.0.1, local=32.0.0
ament_cmake_export_link_flags             : latest=2.8.7, local=2.8.7
ament_cmake_libraries                     : latest=2.8.7, local=2.8.7
ament_index_cpp                           : latest=1.13.3, local=1.13.3
ament_cmake_pytest                        : latest=2.8.7, local=2.8.7
tracetools                                : latest=8.10.2, local=8.10.2
ament_cmake_cpplint                       : latest=0.20.6, local=0.20.6
ament_cmake_export_targets                : latest=2.8.7, local=2.8.7
ament_cmake_gmock                         : latest=2.8.7, local=2.8.7
ament_copyright                           : latest=0.20.6, local=0.20.6
rcl_lifecycle                             : latest=10.4.4, local=10.4.4
rmw_implementation_cmake                  : latest=7.10.1, local=7.10.1
sros2                                      : latest=0.16.6, local=0.16.6
rclcpp                                    : latest=32.0.1, local=32.0.0
diagnostic_msgs                           : latest=5.9.2, local=5.9.2
launch_testing_ament_cmake                : latest=3.9.7, local=3.9.7
rosidl_runtime_c                          : latest=5.2.1, local=5.2.1
rosidl_dynamic_typesupport                : latest=0.4.1, local=0.4.1
rosidl_typesupport_c                      : latest=3.4.2, local=3.4.2
stereo_msgs                               : latest=5.9.2, local=5.9.2
ros2launch                                : latest=0.29.9, local=0.29.8
ament_cmake_include_directories           : latest=2.8.7, local=2.8.7
std_msgs                                  : latest=5.9.2, local=5.9.2
common_interfaces                         : latest=5.9.2, local=5.9.2
rosidl_typesupport_cpp                    : latest=3.4.2, local=3.4.2
rmw_security_common                       : latest=7.10.1, local=7.10.1
builtin_interfaces                        : latest=2.4.5, local=2.4.5
rosidl_default_runtime                    : latest=1.8.1, local=1.8.1
rosidl_generator_cpp                      : latest=5.2.1, local=5.2.1
ament_cmake_export_include_directories    : latest=2.8.7, local=2.8.7
ros2node                                  : latest=0.40.7, local=0.40.7
rosidl_core_generators                    : latest=0.4.3, local=0.4.3
rmw_fastrtps_cpp                          : latest=9.4.8, local=9.4.8
ros2multicast                             : latest=0.40.7, local=0.40.7
ament_cmake_ros                           : latest=0.15.8, local=0.15.8
ament_pep257                              : latest=0.20.6, local=0.20.6
ament_uncrustify                          : latest=0.20.6, local=0.20.6
rosidl_typesupport_interface              : latest=5.2.1, local=5.2.1
rosidl_typesupport_fastrtps_c             : latest=3.9.5, local=3.9.5
rosidl_buffer_py                          : latest=5.2.1, local=5.2.1
rosidl_parser                             : latest=5.2.1, local=5.2.1
ament_cmake_export_dependencies           : latest=2.8.7, local=2.8.7
type_description_interfaces               : latest=2.4.5, local=2.4.5
ament_cpplint                             : latest=0.20.6, local=0.20.6
trajectory_msgs                           : latest=5.9.2, local=5.9.2
rosidl_pycommon                           : latest=5.2.1, local=5.2.1
composition_interfaces                    : latest=2.4.5, local=2.4.5
ament_cmake_cppcheck                      : latest=0.20.6, local=0.20.6
ament_cmake_export_definitions            : latest=2.8.7, local=2.8.7
ament_flake8                              : latest=0.20.6, local=0.20.6
ros2action                                : latest=0.40.7, local=0.40.7
rosidl_runtime_cpp                        : latest=5.2.1, local=5.2.1
rmw_implementation                        : latest=3.1.5, local=3.1.5
ament_cmake                               : latest=2.8.7, local=2.8.7
rosidl_adapter                            : latest=5.2.1, local=5.2.1
unique_identifier_msgs                    : latest=2.8.1, local=2.8.1
rcpputils                                 : latest=2.14.5, local=2.14.5
ros2interface                             : latest=0.40.7, local=0.40.7
ament_cppcheck                            : latest=0.20.6, local=0.20.6
launch_testing                            : latest=3.9.7, local=3.9.7
ament_lint_auto                           : latest=0.20.6, local=0.20.6
rclpy                                     : latest=10.0.10, local=10.0.10
rosgraph_msgs                             : latest=2.4.5, local=2.4.5
rosidl_default_generators                 : latest=1.8.1, local=1.8.1
rosidl_typesupport_fastrtps_cpp           : latest=3.9.5, local=3.9.5
visualization_msgs                        : latest=5.9.2, local=5.9.2
ros2plugin                                : latest=5.8.4, local=5.8.4
rosidl_dynamic_typesupport_fastrtps       : latest=0.5.1, local=0.5.1
ros_workspace                             : latest=1.0.3, local=1.0.3
ament_cmake_xmllint                       : latest=0.20.6, local=0.20.6
libyaml_vendor                            : latest=1.8.1, local=1.8.1
rosidl_typesupport_introspection_c        : latest=5.2.1, local=5.2.1
ament_cmake_lint_cmake                    : latest=0.20.6, local=0.20.6
std_srvs                                  : latest=5.9.2, local=5.9.2
service_msgs                              : latest=2.4.5, local=2.4.5
ros2service                               : latest=0.40.7, local=0.40.7
rosidl_core_runtime                       : latest=0.4.3, local=0.4.3
rclcpp_lifecycle                          : latest=32.0.1, local=32.0.0
ros2run                                   : latest=0.40.7, local=0.40.7
rcl_logging_implementation                : latest=3.4.1, local=3.4.1
ros2pkg                                   : latest=0.40.7, local=0.40.7
rcl_logging_interface                     : latest=3.4.1, local=3.4.1
launch                                    : latest=3.9.7, local=3.9.7
rosidl_buffer_backend_registry            : latest=5.2.1, local=5.2.1
geometry_msgs                             : latest=5.9.2, local=5.9.2
console_bridge_vendor                     : latest=1.9.1, local=1.9.1
spdlog_vendor                             : latest=1.8.0, local=1.8.0
rosidl_generator_c                        : latest=5.2.1, local=5.2.1
rmw_test_fixture_implementation           : latest=0.15.8, local=0.15.8
ament_cmake_auto                          : latest=2.8.7, local=2.8.7
rosidl_buffer                             : latest=5.2.1, local=5.2.1
rmw_dds_common                            : latest=6.0.0, local=6.0.0
ros2lifecycle                             : latest=0.40.7, local=0.40.7
rcl_interfaces                            : latest=2.4.5, local=2.4.5
ament_cmake_pep257                        : latest=0.20.6, local=0.20.6
launch_yaml                               : latest=3.9.7, local=3.9.7
launch_xml                                : latest=3.9.7, local=3.9.7
rcl_action                                : latest=10.4.4, local=10.4.4
shape_msgs                                : latest=5.9.2, local=5.9.2
ament_cmake_flake8                        : latest=0.20.6, local=0.20.6
rcl_yaml_param_parser                     : latest=10.4.4, local=10.4.4
rosidl_cli                                : latest=5.2.1, local=5.2.1
ament_cmake_gen_version_h                 : latest=2.8.7, local=2.8.7
ros2cli                                   : latest=0.40.7, local=0.40.7
rclcpp_components                         : latest=32.0.1, local=32.0.0
ros_environment                           : latest=4.5.1, local=4.5.1
rosidl_buffer_backend                     : latest=5.2.1, local=5.2.1
rmw_fastrtps_shared_cpp                   : latest=9.4.8, local=9.4.8
rosidl_typesupport_introspection_cpp      : latest=5.2.1, local=5.2.1
rcl_logging_spdlog                        : latest=3.4.1, local=3.4.1
rpyutils                                  : latest=0.7.2, local=0.7.2
ament_cmake_copyright                     : latest=0.20.6, local=0.20.6
lifecycle_msgs                            : latest=2.4.5, local=2.4.5
sros2_cmake                               : latest=0.16.6, local=0.16.6
ament_lint                                : latest=0.20.6, local=0.20.6
rosidl_generator_type_description         : latest=5.2.1, local=5.2.1
ros2doctor                                : latest=0.40.7, local=0.40.7
tf2_sensor_msgs                           : latest=0.45.7, local=0.45.7
tf2                                       : latest=0.45.7, local=0.45.7
lz4_cmake_module                          : latest=0.33.3, local=0.33.3
rosbag2_cpp                               : latest=0.33.3, local=0.33.3
ros2bag                                   : latest=0.33.3, local=0.33.3
tf2_py                                    : latest=0.45.7, local=0.45.7
message_filters                           : latest=7.4.2, local=7.4.1
tf2_eigen_kdl                             : latest=0.45.7, local=0.45.7
rosbag2_storage_mcap                      : latest=0.33.3, local=0.33.3
rosbag2_py                                : latest=0.33.3, local=0.33.3
zstd_cmake_module                         : latest=0.33.3, local=0.33.3
tf2_ros                                   : latest=0.45.7, local=0.45.7
kdl_parser                                : latest=3.0.1, local=3.0.1
tf2_tools                                 : latest=0.45.7, local=0.45.7
rosbag2                                   : latest=0.33.3, local=0.33.3
tf2_eigen                                 : latest=0.45.7, local=0.45.7
rosbag2_compression_zstd                  : latest=0.33.3, local=0.33.3
rosbag2_transport                         : latest=0.33.3, local=0.33.3
yaml_cpp_vendor                           : latest=9.2.1, local=9.2.1
robot_state_publisher                     : latest=3.5.5, local=3.5.5
tf2_kdl                                   : latest=0.45.7, local=0.45.7
rosbag2_storage_default_plugins           : latest=0.33.3, local=0.33.3
tf2_ros_py                                : latest=0.45.7, local=0.45.7
tf2_geometry_msgs                         : latest=0.45.7, local=0.45.7
keyboard_handler                          : latest=0.5.1, local=0.5.1
urdf_parser_plugin                        : latest=2.13.2, local=2.13.2
sensor_msgs_py                            : latest=5.9.2, local=5.9.2
ros_base                                  : latest=0.13.0, local=0.13.0
tf2_bullet                                : latest=0.45.7, local=0.45.7
eigen3_cmake_module                       : latest=0.5.1, local=0.5.1
urdf                                      : latest=2.13.2, local=2.13.2
rosbag2_storage                           : latest=0.33.3, local=0.33.3
tf2_msgs                                  : latest=0.45.7, local=0.45.7
mcap_vendor                               : latest=0.33.3, local=0.33.3
rosbag2_storage_sqlite3                   : latest=0.33.3, local=0.33.3
geometry2                                 : latest=0.45.7, local=0.45.7
rosbag2_compression                       : latest=0.33.3, local=0.33.3
rosbag2_interfaces                        : latest=0.33.3, local=0.33.3

   PLATFORM INFORMATION
system           : Linux
platform info    : Linux-6.17.0-35-generic-x86_64-with-glibc2.43
release          : 6.17.0-35-generic
processor        : 

   QOS COMPATIBILITY LIST
compatibility status    : No publisher/subscriber pairs found

   RMW MIDDLEWARE
middleware name    : rmw_fastrtps_cpp

   ROS 2 INFORMATION
distribution name      : lyrical
distribution type      : ros2
distribution status    : active
release platforms      : {'debian': ['trixie'], 'rhel': ['10'], 'ubuntu': ['resolute']}

   SERVICE LIST
service          : none
service count    : 0
client count     : 0

   TOPIC LIST
topic               : none
publisher count     : 0
subscriber count    : 0

```
</details>


### Steps to reproduce issue

- Operating System: Ubuntu 26.04 LTS (Resolute Raccoon)
- ROS 2 Distro: Lyrical (`ros:lyrical` Docker image)
- Version: `ros-lyrical-rclcpp` 32.0.0, `ros-lyrical-rmw-fastrtps-cpp` 9.4.8, `ros-lyrical-rosidl-typesupport-fastrtps-cpp` 3.9.5

### Steps to reproduce issue

A publisher creates 4 publishers with `TRANSIENT_LOCAL` durability QoS (depth 1, reliable) and publishes one message on each, then stays alive (`rclcpp::spin`). A **separate process**, started ~2 seconds later, creates matching `TRANSIENT_LOCAL` subscriptions and waits up to 5 seconds for each topic to deliver its retained message.

Minimal standalone `ament_cmake` package (builds with plain `colcon build`):

<details>
<summary>package.xml</summary>

```xml
<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="http://download.ros.org/schema/package_format3.xsd"?>
<package format="3">
  <name>transient_local_test</name>
  <version>0.0.1</version>
  <description>Minimal reproducer: transient_local QoS publisher/subscriber across 4 message types</description>
  <maintainer email="jane.doe@example.com">Jane Doe</maintainer>
  <license>Apache-2.0</license>

  <buildtool_depend>ament_cmake</buildtool_depend>

  <depend>rclcpp</depend>
  <depend>std_msgs</depend>
  <depend>nav_msgs</depend>

  <export>
    <build_type>ament_cmake</build_type>
  </export>
</package>
```
</details>

<details>
<summary>CMakeLists.txt</summary>

```cmake
cmake_minimum_required(VERSION 3.8)
project(transient_local_test)

if(NOT CMAKE_CXX_STANDARD)
  set(CMAKE_CXX_STANDARD 17)
endif()

find_package(ament_cmake REQUIRED)
find_package(rclcpp REQUIRED)
find_package(std_msgs REQUIRED)
find_package(nav_msgs REQUIRED)

add_executable(transient_publisher src/publisher_node.cpp)
target_link_libraries(transient_publisher
  rclcpp::rclcpp
  std_msgs::std_msgs
  nav_msgs::nav_msgs)

add_executable(transient_subscriber_test src/subscriber_test.cpp)
target_link_libraries(transient_subscriber_test
  rclcpp::rclcpp
  std_msgs::std_msgs
  nav_msgs::nav_msgs)

install(TARGETS
  transient_publisher
  transient_subscriber_test
  DESTINATION lib/${PROJECT_NAME})

ament_package()
```
</details>

<details>
<summary>src/publisher_node.cpp</summary>

```cpp
#include <memory>

#include "nav_msgs/msg/occupancy_grid.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32_multi_array.hpp"
#include "std_msgs/msg/u_int8.hpp"
#include "std_msgs/msg/u_int8_multi_array.hpp"

class TransientPublisher : public rclcpp::Node
{
public:
  TransientPublisher()
  : Node("transient_publisher")
  {
    const auto qos = rclcpp::QoS(1).transient_local().reliable();

    grid_pub_ = create_publisher<nav_msgs::msg::OccupancyGrid>(
      "transient_test/occupancy_grid", qos);
    uint8_array_pub_ = create_publisher<std_msgs::msg::UInt8MultiArray>(
      "transient_test/uint8_multi_array", qos);
    uint8_pub_ = create_publisher<std_msgs::msg::UInt8>(
      "transient_test/uint8", qos);
    float32_array_pub_ = create_publisher<std_msgs::msg::Float32MultiArray>(
      "transient_test/float32_multi_array", qos);

    publish_all();
  }

private:
  void publish_all()
  {
    nav_msgs::msg::OccupancyGrid grid;
    grid.header.stamp = now();
    grid.header.frame_id = "map";
    grid.info.resolution = 1.0f;
    grid.info.width = 2;
    grid.info.height = 2;
    grid.data = {0, 100, -1, 50};
    grid_pub_->publish(grid);

    std_msgs::msg::UInt8MultiArray uint8_array;
    uint8_array.data = {1, 2, 3};
    uint8_array_pub_->publish(uint8_array);

    std_msgs::msg::UInt8 uint8_msg;
    uint8_msg.data = 42;
    uint8_pub_->publish(uint8_msg);

    std_msgs::msg::Float32MultiArray float32_array;
    float32_array.data = {1.1f, 2.2f, 3.3f};
    float32_array_pub_->publish(float32_array);

    RCLCPP_INFO(
      get_logger(),
      "Published all 4 transient_local messages, keeping node alive for late-joining subscribers...");
  }

  rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr grid_pub_;
  rclcpp::Publisher<std_msgs::msg::UInt8MultiArray>::SharedPtr uint8_array_pub_;
  rclcpp::Publisher<std_msgs::msg::UInt8>::SharedPtr uint8_pub_;
  rclcpp::Publisher<std_msgs::msg::Float32MultiArray>::SharedPtr float32_array_pub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TransientPublisher>());
  rclcpp::shutdown();
  return 0;
}
```
</details>

<details>
<summary>src/subscriber_test.cpp</summary>

```cpp
#include <atomic>
#include <chrono>
#include <future>
#include <memory>
#include <string>

#include "nav_msgs/msg/occupancy_grid.hpp"
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32_multi_array.hpp"
#include "std_msgs/msg/u_int8.hpp"
#include "std_msgs/msg/u_int8_multi_array.hpp"

using namespace std::chrono_literals;

class TransientSubscriberTest : public rclcpp::Node
{
public:
  TransientSubscriberTest()
  : Node("transient_subscriber_test")
  {
    const auto qos = rclcpp::QoS(1).transient_local().reliable();

    grid_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
      "transient_test/occupancy_grid", qos,
      [this](const nav_msgs::msg::OccupancyGrid::SharedPtr) {
        if (!grid_received_.exchange(true)) {
          grid_promise_.set_value(true);
        }
      });

    uint8_array_sub_ = create_subscription<std_msgs::msg::UInt8MultiArray>(
      "transient_test/uint8_multi_array", qos,
      [this](const std_msgs::msg::UInt8MultiArray::SharedPtr) {
        if (!uint8_array_received_.exchange(true)) {
          uint8_array_promise_.set_value(true);
        }
      });

    uint8_sub_ = create_subscription<std_msgs::msg::UInt8>(
      "transient_test/uint8", qos,
      [this](const std_msgs::msg::UInt8::SharedPtr) {
        if (!uint8_received_.exchange(true)) {
          uint8_promise_.set_value(true);
        }
      });

    float32_array_sub_ = create_subscription<std_msgs::msg::Float32MultiArray>(
      "transient_test/float32_multi_array", qos,
      [this](const std_msgs::msg::Float32MultiArray::SharedPtr) {
        if (!float32_array_received_.exchange(true)) {
          float32_array_promise_.set_value(true);
        }
      });
  }

  std::promise<bool> grid_promise_;
  std::promise<bool> uint8_array_promise_;
  std::promise<bool> uint8_promise_;
  std::promise<bool> float32_array_promise_;

private:
  std::atomic<bool> grid_received_{false};
  std::atomic<bool> uint8_array_received_{false};
  std::atomic<bool> uint8_received_{false};
  std::atomic<bool> float32_array_received_{false};

  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr grid_sub_;
  rclcpp::Subscription<std_msgs::msg::UInt8MultiArray>::SharedPtr uint8_array_sub_;
  rclcpp::Subscription<std_msgs::msg::UInt8>::SharedPtr uint8_sub_;
  rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr float32_array_sub_;
};

namespace
{

bool wait_for(
  rclcpp::executors::SingleThreadedExecutor & executor,
  std::future<bool> & future,
  const std::string & label)
{
  const auto result = executor.spin_until_future_complete(future, 5s);
  const bool ok = (result == rclcpp::FutureReturnCode::SUCCESS);
  RCLCPP_INFO(
    rclcpp::get_logger("transient_subscriber_test"), "%s: %s",
    label.c_str(), ok ? "PASS" : "FAIL (timeout after 5s)");
  return ok;
}

}  // namespace

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);

  auto node = std::make_shared<TransientSubscriberTest>();
  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);

  auto grid_future = node->grid_promise_.get_future();
  auto uint8_array_future = node->uint8_array_promise_.get_future();
  auto uint8_future = node->uint8_promise_.get_future();
  auto float32_array_future = node->float32_array_promise_.get_future();

  bool all_ok = true;
  all_ok = wait_for(executor, grid_future, "nav_msgs/OccupancyGrid       ") && all_ok;
  all_ok = wait_for(executor, uint8_array_future, "std_msgs/UInt8MultiArray     ") && all_ok;
  all_ok = wait_for(executor, uint8_future, "std_msgs/UInt8               ") && all_ok;
  all_ok = wait_for(executor, float32_array_future, "std_msgs/Float32MultiArray   ") && all_ok;

  rclcpp::shutdown();

  if (all_ok) {
    RCLCPP_INFO(rclcpp::get_logger("transient_subscriber_test"), "ALL TESTS PASSED");
    return 0;
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("transient_subscriber_test"), "SOME TESTS FAILED");
    return 1;
  }
}
```
</details>

Build and run (inside `docker run -it --rm ros:lyrical`, after mounting a workspace containing this package under `src/`):

```bash
source /opt/ros/lyrical/setup.bash
colcon build --packages-select transient_local_test
source install/setup.bash

./install/transient_local_test/lib/transient_local_test/transient_publisher &
sleep 2
./install/transient_local_test/lib/transient_local_test/transient_subscriber_test
```


### Expected behavior

All 4 topics should deliver their retained message to the late-joining subscriber within the 5 second timeout, since all 4 publishers use identical `TRANSIENT_LOCAL` durability QoS and the publisher process is still alive:

```
nav_msgs/OccupancyGrid       : PASS
std_msgs/UInt8MultiArray     : PASS
std_msgs/UInt8               : PASS
std_msgs/Float32MultiArray   : PASS
ALL TESTS PASSED
```

### Actual behavior

`std_msgs/msg/UInt8MultiArray` never reaches the late-joining subscriber and times out after 5 seconds, while the other three types (including `std_msgs/msg/Float32MultiArray`, which has the identical `MultiArrayLayout` + array-of-primitives structure) are delivered immediately:

```
[INFO] [transient_publisher]: Published all 4 transient_local messages, keeping node alive for late-joining subscribers...
[INFO] [transient_subscriber_test]: nav_msgs/OccupancyGrid       : PASS
[INFO] [transient_subscriber_test]: std_msgs/UInt8MultiArray     : FAIL (timeout after 5s)
[INFO] [transient_subscriber_test]: std_msgs/UInt8               : PASS
[INFO] [transient_subscriber_test]: std_msgs/Float32MultiArray   : PASS
[ERROR] [transient_subscriber_test]: SOME TESTS FAILED
```

Reproduced consistently across 3/3 runs.

**This is a regression specific to Lyrical.** The identical test (rebuilt from source, no code changes) was run against `ros:kilted` and `ros:jazzy` and passes 4/4 on both:

| Distro | OccupancyGrid | UInt8MultiArray | UInt8 | Float32MultiArray |
|---|---|---|---|---|
| `ros:jazzy` (LTS) | PASS | PASS | PASS | PASS |
| `ros:kilted` | PASS | PASS | PASS | PASS |
| `ros:lyrical` | PASS | **FAIL** | PASS | PASS |

**Independently reproduced in `rclpy`** (not just `rclcpp`), ruling out a C++/rclcpp-specific bug:

```python
# publisher (kept alive)
qos = QoSProfile(depth=1, durability=DurabilityPolicy.TRANSIENT_LOCAL, reliability=ReliabilityPolicy.RELIABLE)
pub_u8 = node.create_publisher(UInt8MultiArray, 'transient_test/uint8_multi_array_py', qos)
pub_u8.publish(UInt8MultiArray(data=[1, 2, 3]))

# subscriber, started 2s later
node.create_subscription(UInt8MultiArray, 'transient_test/uint8_multi_array_py', cb_u8, qos)
# -> callback never fires within 5s; Float32MultiArray on an equivalent topic/QoS does fire
```


### Additional information

### Suspected cause

Since the failure is specific to `uint8[]` (IDL `octet` sequence) fields and does **not** reproduce for `float32[]` with the same `MultiArrayLayout` wrapper, this points at `rmw_fastrtps_cpp` / `rosidl_typesupport_fastrtps_cpp` mishandling octet-sequence fields in the `TRANSIENT_LOCAL` durability history cache (e.g. a serialization or zero-copy/loan fast-path for byte arrays that bypasses the durability cache), rather than a generic QoS/durability plumbing issue. `rmw_cyclonedds_cpp` or `rmw_zenoh_cpp` are not affected.

the repro above is fully self-contained and takes <30s to build and run.

--- a/rmw_fastrtps_cpp/src/publisher.cpp
+++ b/rmw_fastrtps_cpp/src/publisher.cpp
@@ -255,10 +255,24 @@ rmw_fastrtps_cpp::create_publisher(
     writer_qos.data_sharing().off();
   }
 
-  // Detect buffer-aware message type
-  bool has_buffer_fields = callbacks->has_buffer_fields;
+  // Resolve the effective QoS before deciding whether to enable buffer endpoints.
+  const bool has_buffer_fields = callbacks->has_buffer_fields;
+  if (!get_datawriter_qos(
+      *qos_policies, *type_supports->get_type_hash_func(type_supports),
+      writer_qos, nullptr, nullptr))
+  {
+    RMW_SET_ERROR_MSG("create_publisher() failed setting data writer QoS");
+    return nullptr;
+  }
+
+  // Buffer backends do not support transient-local durability. Those publishers
+  // use only the base DataWriter, which provides the required history cache.
+  const bool use_buffer_endpoints = has_buffer_fields &&
+    writer_qos.durability().kind !=
+    eprosima::fastdds::dds::TRANSIENT_LOCAL_DURABILITY_QOS;
+
   std::unordered_map<std::string, std::string> backend_metadata;
-  if (has_buffer_fields) {
+  if (use_buffer_endpoints) {
     auto * backend_context =
       static_cast<rmw_fastrtps_cpp::BufferBackendContext *>(
       participant_info->buffer_serialization_context_);
@@ -271,16 +285,15 @@ rmw_fastrtps_cpp::create_publisher(
     if (backend_metadata.find("cpu") == backend_metadata.end()) {
       backend_metadata["cpu"] = "";
     }
-  }
 
-  // Get QoS from RMW, optionally encoding buffer backend info in user_data
-  if (!get_datawriter_qos(
-      *qos_policies, *type_supports->get_type_hash_func(type_supports),
-      writer_qos, nullptr,
-      has_buffer_fields ? &backend_metadata : nullptr))
-  {
-    RMW_SET_ERROR_MSG("create_publisher() failed setting data writer QoS");
-    return nullptr;
+    // Rebuild user_data with buffer backend metadata for discovery.
+    if (!get_datawriter_qos(
+        *qos_policies, *type_supports->get_type_hash_func(type_supports),
+        writer_qos, nullptr, &backend_metadata))
+    {
+      RMW_SET_ERROR_MSG("create_publisher() failed setting buffer-aware data writer QoS");
+      return nullptr;
+    }
   }
 
   // Apply resource limits QoS if the type is keyed
@@ -348,8 +361,8 @@ rmw_fastrtps_cpp::create_publisher(
   rmw_publisher->options = *publisher_options;
 
   // Buffer-aware publisher setup
-  info->is_buffer_aware_ = has_buffer_fields;
-  if (has_buffer_fields) {
+  info->is_buffer_aware_ = use_buffer_endpoints;
+  if (use_buffer_endpoints) {
     info->serialization_context_ = participant_info->buffer_serialization_context_;
     info->backend_metadata_ = backend_metadata;
     info->participant_ = dds_participant;

--- a/rmw_fastrtps_cpp/src/rmw_publish.cpp
+++ b/rmw_fastrtps_cpp/src/rmw_publish.cpp
@@ -250,7 +250,12 @@ rmw_publish(
     return RMW_RET_INVALID_ARGUMENT);
 
   auto info = static_cast<CustomPublisherInfo *>(publisher->data);
-  if (info->is_buffer_aware_) {
+  // Route transient-local publishers through the main DataWriter below as
+  // transient-local is not supported by buffer backends.
+  if (info->is_buffer_aware_ &&
+    info->data_writer_->get_qos().durability().kind !=
+    eprosima::fastdds::dds::TRANSIENT_LOCAL_DURABILITY_QOS)
+  {
     // Check whether any non-backend-aware (legacy) subscribers exist.
     // If so, skip buffer channels entirely and fall through to the main
     // DataWriter so old RMW endpoints receive the message.  Buffer-aware
@@ -268,9 +273,6 @@ rmw_publish(
       publish_to_buffer_endpoints(info, ros_message, publisher);
       return RMW_RET_OK;
     }
-    // Legacy subscribers present — publish via main (legacy) DataWriter.
-    return rmw_fastrtps_shared_cpp::__rmw_publish(
-      eprosima_fastrtps_identifier, publisher, ros_message, allocation);
   }
 
   return rmw_fastrtps_shared_cpp::__rmw_publish(

--- a/rmw_fastrtps_cpp/src/rmw_publish.cpp
+++ b/rmw_fastrtps_cpp/src/rmw_publish.cpp
@@ -250,12 +250,7 @@ rmw_publish(
     return RMW_RET_INVALID_ARGUMENT);
 
   auto info = static_cast<CustomPublisherInfo *>(publisher->data);
-  // Route transient-local publishers through the main DataWriter below as
-  // transient-local is not supported by buffer backends.
-  if (info->is_buffer_aware_ &&
-    info->data_writer_->get_qos().durability().kind !=
-    eprosima::fastdds::dds::TRANSIENT_LOCAL_DURABILITY_QOS)
-  {
+  if (info->is_buffer_aware_) {
     // Check whether any non-backend-aware (legacy) subscribers exist.
     // If so, skip buffer channels entirely and fall through to the main
     // DataWriter so old RMW endpoints receive the message.  Buffer-aware

--- a/rmw_fastrtps_cpp/src/subscription.cpp
+++ b/rmw_fastrtps_cpp/src/subscription.cpp
@@ -695,13 +695,27 @@ __create_subscription(
     reader_qos.data_sharing().off();
   }
 
-  // Detect buffer-aware message type
-  bool has_buffer_fields = callbacks->has_buffer_fields;
+  // Resolve the effective QoS before deciding whether to enable buffer endpoints.
+  const bool has_buffer_fields = callbacks->has_buffer_fields;
+  if (!get_datareader_qos(
+      *qos_policies, *type_supports->get_type_hash_func(type_supports),
+      reader_qos, nullptr, nullptr))
+  {
+    RMW_SET_ERROR_MSG("create_subscription() failed setting data reader QoS");
+    return nullptr;
+  }
+
+  // A transient-local subscription can only match publishers that use the base
+  // DataWriter, so its custom buffer endpoints would never receive data.
+  const bool use_buffer_endpoints = has_buffer_fields &&
+    reader_qos.durability().kind !=
+    eprosima::fastdds::dds::TRANSIENT_LOCAL_DURABILITY_QOS;
+
   std::unordered_map<std::string, std::string> filtered_backends;
   std::vector<std::string> my_backend_types;
   bool cpu_only = false;
 
-  if (has_buffer_fields) {
+  if (use_buffer_endpoints) {
     std::unordered_map<std::string, std::string> all_backends;
     auto * backend_context =
       static_cast<rmw_fastrtps_cpp::BufferBackendContext *>(
@@ -764,15 +778,15 @@ __create_subscription(
         }
       }
     }
-  }
 
-  if (!get_datareader_qos(
-      *qos_policies, *type_supports->get_type_hash_func(type_supports),
-      reader_qos, nullptr,
-      has_buffer_fields ? &filtered_backends : nullptr))
-  {
-    RMW_SET_ERROR_MSG("create_subscription() failed setting data reader QoS");
-    return nullptr;
+    // Rebuild user_data with buffer backend metadata for discovery.
+    if (!get_datareader_qos(
+        *qos_policies, *type_supports->get_type_hash_func(type_supports),
+        reader_qos, nullptr, &filtered_backends))
+    {
+      RMW_SET_ERROR_MSG("create_subscription() failed setting buffer-aware data reader QoS");
+      return nullptr;
+    }
   }
 
   // Apply resource limits QoS if the type is keyed
@@ -844,9 +858,9 @@ __create_subscription(
   rmw_subscription->is_cft_supported = true;
 
   // Buffer-aware subscription setup
-  info->is_buffer_aware_ = has_buffer_fields;
-  info->is_cpu_only_ = has_buffer_fields && cpu_only;
-  if (has_buffer_fields) {
+  info->is_buffer_aware_ = use_buffer_endpoints;
+  info->is_cpu_only_ = use_buffer_endpoints && cpu_only;
+  if (use_buffer_endpoints) {
     info->serialization_context_ = participant_info->buffer_serialization_context_;
     info->my_backend_types_ = std::move(my_backend_types);
     info->local_endpoint_info_ = rmw_get_zero_initialized_topic_endpoint_info();

